sam/feature_engineering/base_feature_engineering.py

```diff
@@ -0,0 +1,102 @@
+import logging
+from abc import ABC, abstractmethod
+from typing import Callable
+
+import pandas as pd
+from sklearn.base import TransformerMixin
+
+
+class BaseFeatureEngineer(TransformerMixin, ABC):
+    """
+    Base class for feature engineering.
+    To use this class, you need to implement the feature_engineer method.
+    """
+
+    def __init__(self):
+        pass
+
+    @abstractmethod
+    def feature_engineer_(self, X, y=None) -> pd.DataFrame:
+        raise NotImplementedError("You need to implement the feature_engineer method.")
+
+    def fit(self, X, y=None):
+        self._feature_names = self.feature_engineer_(X, y).columns.tolist()
+        return self
+
+    def transform(self, X, y=None) -> pd.DataFrame:
+        logging.info("Feature engineering - input shape: %s", X.shape)
+        X_out = self.feature_engineer_(X, y)
+        logging.info("Feature engineering - output shape: %s", X_out.shape)
+        return X_out
+
+    def get_feature_names(self) -> list:
+        """
+        Function for obtaining feature names. Generally used instead of the attribute, and more
+        compatible with the sklearn API.
+
+        Returns
+        -------
+        list:
+            list of feature names
+        """
+        # check_is_fitted(self.feature_engineer_, "_feature_names")
+        return self._feature_names
+
+
+class FeatureEngineer(BaseFeatureEngineer):
+    """
+    Feature engineering class. This class is used to feature engineer the data using default
+    methods. You can implement your own feature engineering code and use the `_from_function`
```

Comment:
Where is this `_from_function` defined?

Reply:
Nowhere, I tried a static method but didn't work out. Changed the docs.